tests/integration/backward_compatible/heartbeat_test.py

```diff
@@ -48,25 +53,57 @@ def collector(c, *_):
+            self.client, addr, 2, start_simulation, assertion_succeeded
+        )
 
         def assert_heartbeat_stopped_and_restored():
-            self.assertEqual(1, len(connection_added_collector.connections))
-            self.assertEqual(1, len(connection_removed_collector.connections))
-            stopped_connection = connection_added_collector.connections[0]
-            restored_connection = connection_removed_collector.connections[0]
+            self.assertTrue(len(connection_added_collector.connections) >= 1)
+            self.assertTrue(len(connection_removed_collector.connections) >= 1)
+
+            stopped_connection = connection_removed_collector.connections[0]
+            restored_connection = connection_added_collector.connections[0]
+
             self.assertEqual(
-                stopped_connection.connected_address, Address(member2.host, member2.port)
+                stopped_connection.connected_address,
+                Address(member2.host, member2.port),
             )
             self.assertEqual(
-                restored_connection.connected_address, Address(member2.host, member2.port)
+                restored_connection.connected_address,
+                Address(member2.host, member2.port),
             )
+            assertion_succeeded[0] = True
 
+        start_simulation.set()
         self.assertTrueEventually(assert_heartbeat_stopped_and_restored)
+        simulation_thread.join()
 
     @staticmethod
-    def simulate_heartbeat_lost(client, address, timeout):
-        for connection in client._connection_manager.active_connections.values():
-            if connection.remote_address == address:
-                connection.last_read_time -= timeout
-                break
+    def simulate_heartbeat_loss(client, address, timeout, start_simulation, assertion_succeeded):
+        def run():
+            # Wait until the main thread signals we should start the simulation
+            start_simulation.wait()
+
+            # It is possible for client to override the set last_read_time
+            # of the connection, in case of the periodically sent heartbeat
+            # requests getting responses, right after we try to set a new
+            # value to it, before the next iteration of the heartbeat manager.
+            # In this case, the connection won't be closed, and the test would
+            # fail. To avoid it, we will try multiple times.
+            for i in range(10):
+                if assertion_succeeded[0]:
+                    # We have successfully simulated heartbeat loss
+                    return
+
+                for connection in client._connection_manager.active_connections.values():
+                    if connection.remote_address == address:
+                        connection.last_read_time -= timeout
+                        break
+
+                time.sleep((i + 1) * 0.1)
+
+        t = threading.Thread(target=run)
```

Comment:
Ok, the logic seems very similar to my fix to HeartbeatFromServerTest in the Node.js client, but why do we need a thread and threading.Event here? Looks like you want the thread execution to wait something, but the event flag is set immediately after thread is started. So, it actually does not wait for anything in other threads, does it?

Reply:
It is a minor thing, but I wanted the loop to start right before the assertion. I could remove it if you want, and it probably will not affect anything

Reply:
simulate_heartbeat_loss was starting right before the assertion before this change as well. Wasn't it? I think if threading is not necessary, it can be removed to make code less complex.

Reply:
Sorry looks like threading is necessary here but event can be removed. Right? Then, nevermind. It is minor thing

Reply:
Yes. BTW, I have removed it nevertheless to make it simpler